calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php

```diff
@@ -211,7 +211,19 @@
 			'inputType'               => 'textarea',
 			'eval'                    => array('style'=>'height:60px', 'tl_class'=>'clr'),
 			'sql'                     => "text NULL"
-		)
+		),
+		'imgSize' => array
+		(
+			'exclude'                 => true,
+			'inputType'               => 'imageSize',
+			'reference'               => &$GLOBALS['TL_LANG']['MSC'],
+			'eval'                    => array('rgxp'=>'natural', 'includeBlankOption'=>true, 'nospace'=>true, 'helpwizard'=>true, 'tl_class'=>'w50'),
+			'options_callback' => static function ()
+			{
+				return System::getContainer()->get('contao.image.image_sizes')->getOptionsForUser(BackendUser::getInstance());
+			},
```

Comment:
What about putting this in a class, tagging it with…

```
@Callback(table="tl_calendar_feed ", target="fields.imgSize.options")
@Callback(table="tl_news_feed ", target="fields.imgSize.options")
```
… and using proper DI for both instances?

Reply:
I really dislike the tagging pattern for select fields: It completely removes any reference to the options away from the field definition: You look at the configuration and have no clue a) which options it has or b) even where these are compiled. You have to find the implementation manually somewhere. I would prefer defining a callabe over tagging rather than an anonymous function. 

However, if this it where Contao wants to go I will bow to that.

Reply:
You are definitely right that the loss of reference between an option field and its options is inconvenient.

The notation of a callable / service call can at least provide a hint on where to look for the options source (e.g. `contao.image.image_sizes`), but that still requires a manual lookup of the service ID and a public service - so it's not perfect either.

For the specific use-case of image sizes, there probably should be an easier way to get the options for the current backend user. Either the `ImageSizes` service should know about the current request and optionally infer the current backend user from it or a more general `BackendUserSettings` service should be added which can be queried for options and settings of the current backend user.

Alternatively, since the `ImageSize` widget is already querying the available options from the image sizes service in its validator: Maybe this logic can also be used as a default so you don't have to define the options for "regular" image size options at all?

However, as the current one-line callback to get the image sizes is used in multiple places already, I think it should be used here, too (for now, and maybe get refactored later together with similar callbacks).

Reply:
As a side note: In my opinion, the specification of a closure as a callback should be completely discouraged in the long term, as these can only be defined within PHP code. A conversion / dumping of a DCA into another format - for example YAML - as a result would be impossible.

So that's definitely something to think about in the future.

Reply:
>  but that still requires a manual lookup of the service ID and a public service - so it's not perfect either.

Yes, true – not perfect, but at least I have a service ID to search for in my IDE.

> However, as the current one-line callback to get the image sizes is used in multiple places already, I think it should be used here, too (for now, and maybe get refactored later together with similar callbacks).

That is what I think, too. 

> As a side note: In my opinion, the specification of a closure as a callback should be completely discouraged in the long term, as these can only be defined within PHP code. A conversion / dumping of a DCA into another format - for example YAML - as a result would be impossible.
I feel like the trend goes back [from YAML/XML to  PHP-based config](https://github.com/symfony/symfony/pull/36778) – at least partially. But I agree that anonymous functions in DCAs in general are difficult.

Reply:
I have created a separate issue for this: #3907

Reply:
I am currently refactoring this: I implement a general purpose listener that is used in all other places in the core, too (`tl_layout`, `tl_module`, `tl_content`). I then use this in `tl_news` and `tl_news_feed` as well as `tl_calendar_event`  and `tl_calendar_feed`.